### ISSUE TITLE
Conflict with Divi theme / page builder

### ISSUE THREAD

**Issue by ZebulanStanphill** [xwp/pwa-wp#608]
When using [Divi](https://www.elegantthemes.com/gallery/divi/) with this plugin, visiting any page with wp-admin code loaded will throw this fatal error in the site's PHP error log:

```
PHP Fatal error: Uncaught Error: Undefined constant "ET_BUILDER_PRODUCT_VERSION" in /home/MY_SITE/public_html/wp-content/themes/Divi/includes/builder/feature/Library.php:1141
Stack trace:
#0 /home/MY_SITE/public_html/wp-includes/class-wp-hook.php(303): ET_Builder_Library->wp_hook_admin_enqueue_scripts('service-worker')
#1 /home/MY_SITE/public_html/wp-includes/class-wp-hook.php(327): WP_Hook->apply_filters(NULL, Array)
#2 /home/MY_SITE/public_html/wp-includes/plugin.php(470): WP_Hook->do_action(Array)
#3 /home/MY_SITE/public_html/wp-content/plugins/pwa/wp-includes/class-wp-service-workers.php(168): do_action('admin_enqueue_s...', 'service-worker')
#4 /home/MY_SITE/public_html/wp-content/plugins/pwa/wp-includes/service-workers.php(219): WP_Service_Workers->serve_request()
#5 /home/MY_SITE/public_html/wp-includes/class-wp-hook.php(303): wp_ajax_wp_service_worker('')
#6 /home/MY_SITE/public_html/wp-includes/class-wp-hook.php(327): WP_Hook->apply_filters('', Array)
#7 /home/MY_SITE/public_html/wp-includes/plugin.php(470): WP_Hook->do_action(Array)
#8 /home/MY_SITE/public_html/wp-admin/admin-ajax.php(187): do_action('wp_ajax_wp_serv...')
#9 {main}
thrown in /home/MY_SITE/public_html/wp-content/themes/Divi/includes/builder/feature/Library.php on line 1141
```

And the following errors get logged in the browser dev console:

```
A bad HTTP response code (500) was received when fetching the script.
plugins.php:1
```

```
Uncaught (in promise) TypeError: Failed to register a ServiceWorker for scope ('https://MYSITE/wp-admin/') with script ('https://MYSITE/wp-admin/admin-ajax.php?action=wp_service_worker'): A bad HTTP response code (500) was received when fetching the script.
```

Default WP themes like Twenty Twenty do not have this issue. This issue occurs even when all other plugins are disabled.

It's also worth noting that the errors are only thrown when on an admin page or when viewing a front-end page while logged in. If viewing the front-end while logged out, no errors occur.

Additional environment info:

WordPress 5.8
PHP 8
Divi 4.10.5
PWA 0.6.0

This issue was reported this issue to Divi via their support channel. Their latest response:

<blockquote><p>Thank you so much for your patience so far. We are experiencing quite high chat volume these days. 😊
If that issue only happens with the PWA plugin then you need to contact the plugin developer. Generally, we can't guarantee support with all plugins as we have no control over how they are coded and there are too many plugins created (tens of thousands) to test them all. That being said, most plugins should work assuming they are coded well and no conflicts arise.</p>
<p>﻿But if they did, I recommend you to get in touch with the plugin support so they can check the conflict and provide the fix. 👌</p>
<p>﻿Thank you so much for understanding!</p></blockquote>

**Comment by westonruter** [comment]
Since Divi is a commercial plugin and the source code is not available, there's nothing really I can do. It's something Divi should probably fix. The error is coming from their product after all.

**Comment by ZebulanStanphill** [comment]
@westonruter Technically, Divi is GPL-licensed, so I can provide the [distributed code](https://1drv.ms/u/s!Aj3QiPw9uiwSjNkTlYNCpcb8aSQRYw?e=S0rRBb), although it appears that the JS files are all minified so good luck trying to debug those parts. Although come to think of it, I think the GPL license means I have the legal right to request the unminified source code, so that could help?

But yes, I agree it is probably a bug on their end rather than this plugin. I've informed their support team of your response and I'll let you know what they say.

**Comment by ZebulanStanphill** [comment]
Ended up having to solve the issue on my own. Thankfully, the issue was in the PHP code (which isn't minified), and the fix was literally just one line.

In `includes/builder/feature/Library.php`:

```php
public function wp_hook_admin_enqueue_scripts( $page ) {
	global $typenow;
	
	require_once ET_BUILDER_DIR . '/functions.php'; // <-- Add this line.
```

I've contacted Divi's support team again to notify them about this fix. Hopefully, it will be included in a future release of the theme/plugin. Until then, hopefully this post will help anyone else running into this problem.
